File: backendPython/utils/receitas_utils.py
import unicodedata
from functools import lru_cache

@lru_cache(maxsize=1024)
def normalizar_texto(texto: str) -> str:
    """Normaliza texto removendo acentos e convertendo para minúsculas"""
    if not texto:
        return ""
    texto = unicodedata.normalize('NFD', texto)
    texto = ''.join(c for c in texto if unicodedata.category(c) != 'Mn')
    return texto.lower()

def buscar_receitas(conn, ingredientes):
    """Busca receitas que contenham todos os ingredientes especificados"""
    if not conn or not ingredientes:
        return []
    
    try:
        cursor = conn.cursor()
        cursor.execute("""
            SELECT r.id, r.titulo, r.nota, r.avaliacoes, r.autor, 
                   r.tempo_preparo, r.link, r.imagem, GROUP_CONCAT(i.item, '|||') as ings
            FROM recipes r
            LEFT JOIN ingredients i ON i.recipe_id = r.id
            GROUP BY r.id
        """)
        
        search_ings = [normalizar_texto(ing.strip()) for ing in ingredientes]
        matches = []
        
        for row in cursor.fetchall():
            row_dict = dict(row)
            all_ings = row_dict.get('ings', '')
            
            if not all_ings:
                continue
            
            recipe_ings = [normalizar_texto(ing) for ing in all_ings.split('|||')]
            
            # Verifica se todos os ingredientes de busca estão presentes
            if all(any(si in ri for ri in recipe_ings) for si in search_ings):
                matches.append({
                    'id': row_dict['id'],
                    'titulo': row_dict['titulo'],
                    'nota': row_dict['nota'],
                    'avaliacoes': row_dict['avaliacoes'],
                    'autor': row_dict['autor'],
                    'tempo_preparo': row_dict['tempo_preparo'],
                    'link': row_dict['link'],
                    'imagem': row_dict['imagem']
                })
        
        # Ordena por nota e avaliações
        def calcular_score(r):
            nota = 0.0
            if r['nota'] and r['nota'] != 'N/A':
                try:
                    nota = float(r['nota'])
                except:
                    pass
            
            aval = 0
            if r['avaliacoes'] and r['avaliacoes'] != 'N/A':
                try:
                    aval = int(r['avaliacoes'].replace(' votos', '').replace(' voto', '').strip())
                except:
                    pass
            
            return (-nota, -aval)
        
        matches.sort(key=calcular_score)
        return matches
        
    except Exception as e:
        print(f"Search Error: {e}")
        return []

File: backendNode/src/utils/receitasUtils.js
/**
 * Normaliza texto removendo acentos e convertendo para minúsculas
 */
function normalizarTexto(texto) {
  if (!texto) return '';
  
  return texto
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase();
}

/**
 * Busca receitas que contenham todos os ingredientes especificados
 */
async function buscarReceitas(db, ingredientes) {
  if (!db || !ingredientes || ingredientes.length === 0) {
    return [];
  }

  try {
    const query = `
      SELECT r.id, r.titulo, r.nota, r.avaliacoes, r.autor, 
             r.tempo_preparo, r.link, r.imagem, GROUP_CONCAT(i.item, '|||') as ings
      FROM recipes r
      LEFT JOIN ingredients i ON i.recipe_id = r.id
      GROUP BY r.id
    `;

    const rows = await new Promise((resolve, reject) => {
      db.all(query, (err, rows) => {
        if (err) reject(err);
        else resolve(rows || []);
      });
    });

    const searchIngs = ingredientes.map(ing => normalizarTexto(ing.trim()));
    const matches = [];

    for (const row of rows) {
      const allIngs = row.ings || '';

      if (!allIngs) continue;

      const recipeIngs = allIngs.split('|||').map(ing => normalizarTexto(ing));

      // Verifica se todos os ingredientes de busca estão presentes
      const allPresent = searchIngs.every(si => 
        recipeIngs.some(ri => ri.includes(si))
      );

      if (allPresent) {
        matches.push({
          id: row.id,
          titulo: row.titulo,
          nota: row.nota,
          avaliacoes: row.avaliacoes,
          autor: row.autor,
          tempo_preparo: row.tempo_preparo,
          link: row.link,
          imagem: row.imagem
        });
      }
    }

    // Ordena por nota e avaliações
    matches.sort((a, b) => {
      let notaA = 0;
      if (a.nota && a.nota !== 'N/A') {
        try {
          notaA = parseFloat(a.nota);
        } catch (e) {}
      }

      let avalA = 0;
      if (a.avaliacoes && a.avaliacoes !== 'N/A') {
        try {
          avalA = parseInt(a.avaliacoes.replace(' votos', '').replace(' voto', '').trim());
        } catch (e) {}
      }

      let notaB = 0;
      if (b.nota && b.nota !== 'N/A') {
        try {
          notaB = parseFloat(b.nota);
        } catch (e) {}
      }

      let avalB = 0;
      if (b.avaliacoes && b.avaliacoes !== 'N/A') {
        try {
          avalB = parseInt(b.avaliacoes.replace(' votos', '').replace(' voto', '').trim());
        } catch (e) {}
      }

      if (notaB !== notaA) return notaB - notaA;
      return avalB - avalA;
    });

    return matches;
  } catch (error) {
    console.error('Search Error:', error);
    return [];
  }
}

module.exports = {
  normalizarTexto,
  buscarReceitas
};
